File: OOP/6_multilevel_inheritance.cpp
#include <iostream>
#include <string>
using namespace std;

class GrandFather{
    protected:
        string grandfather_name;
    public:
        GrandFather(string grandfather_name){
            this->grandfather_name = grandfather_name;
        };
        void show_grandfather(){
            cout<<grandfather_name<<" is your grandpa."<<endl;
        };
};

class Father: public GrandFather{
    protected:
        string father_name;
    public:
        Father(string grandfather_name,string father_name): 
        GrandFather(grandfather_name){
            this->father_name = father_name;
        }
        void show_father(){
            cout<<father_name<<" is your father."<<endl;
        }
};

class Child: public Father{
    protected:
        string child_name;
    public:
        Child(string grandfather_name,string father_name,string child_name):
        Father(grandfather_name,father_name){
            this->child_name = child_name;
        }

        void show_name(){
            cout<<"Assalamu Alaikum, "<<child_name<<"."<<endl;
        }
};

int main(){
    Child obj1("John","Doe","Asif");
    obj1.show_name();    
    obj1.show_father();    
    obj1.show_grandfather();    
}

File: OOP/6_multilevel_inheritance.py
class Grandfather:
    def __init__(self,grandfather_name) -> None:
        self.grand_father = grandfather_name
    def show_grandfather(self):
        print(f'{self.grand_father} is your Grandfather.')
        
class Father(Grandfather):
    def __init__(self, grandfather_name,father_name) -> None:
        super().__init__(grandfather_name)
        self.father_name = father_name
    def show_father(self):
        print(f'{self.father_name} is your Father.')

class Child(Father):
    def __init__(self, grandfather_name, father_name,child_name) -> None:
        super().__init__(grandfather_name, father_name)
        self.child_name = child_name
    def show_name(self):
        print(f'Assalamu Alaikum, {self.child_name}.')

# here father class inherited from grandfather class and 
# child class inherited from father class.
# Thus,this is an example of multilevel inheritance

obj1 = Child("John","Doe","Asif")
obj1.show_name()
obj1.show_father()
obj1.show_grandfather()